### DISCUSSION TITLE
How to Calculate Grid Density for Given kspacing in Pymatgen?

### DISCUSSION THREAD

**Discussion by hongyi-zhao** [materialsproject/pymatgen#3709] [General]
Hi Pymatgen Team,

I'm currently working on setting up my VASP calculations and trying to understand how to derive the k-point grid size based on the `KSPACING` parameter. Could you guide me on how to calculate or estimate the grid density from a given `KSPACING` value using Pymatgen?

Any advice or pointers to relevant documentation would be greatly appreciated.

Thanks, 
Zhao

**Comment by janosh** [comment]
not sure i understand the question but perhaps the VASP `KSPACING` docs can help https://www.vasp.at/wiki/index.php/KSPACING

using `KSPACING` means VASP places the points equidistant in reciprocal space, i.e. the units are inverse Angstrom. e.g. `kspacing = 0.25` corresponds to `reciprocal_density=64`, meaning 64 kpoints/A^3

**Reply by hongyi-zhao** [reply]
Say, I set the following in `INCAR`:

`KSPACING = 0.02  in \AA ^-1`

Then, I want to know the corresponding [`n1 * n2 * n3` mesh](https://www.vasp.at/wiki/index.php/KPOINTS#Regular_k-point_mesh) when using a `KPOINTS` file.

**Reply by janosh** [reply]
that depends on the shape of the cell in reciprocal space. not sure if `pymatgen` has functionality to auto-convert between these formats for a given `Structure`. if not, PRs would be welcome!

`KSPACING = 0.02` seems very tight. i doubt you need that much

**Reply by hongyi-zhao** [reply]
> that depends on the shape of the cell in reciprocal space. not sure if pymatgen has functionality to auto-convert between these formats for a given Structure. if not, PRs would be welcome!

Based on the specification described in INCAR tag [KSPACING](https://www.vasp.at/wiki/index.php/KSPACING) and source code file `mkpoints.F`, we need to focus on the logic that computes `NKPX`, `NKPY`, and `NKPZ` using the lattice norms (`BNORM`) and the spacing (`KSPACING`). The optional multiplier (`MULTIPL`) is also considered if provided. Here's the implementation in Python:
```python
import math

def calculate_grid_dimensions(bnorm, spacing, multipl=None):
    """
    Calculate the Monkhorst-Pack grid dimensions based on lattice norms and spacing.

    :param bnorm: Lattice norms (list or tuple of three floats).
    :param spacing: KPOINTS spacing (float).
    :param multipl: Optional multiplier (list or tuple of three integers).
    :return: Tuple of grid dimensions (NKPX, NKPY, NKPZ).
    """
    # Calculate initial grid dimensions based on spacing
    nkpx = max(1, math.ceil(bnorm[0] * math.pi * 2 / spacing))
    nkpy = max(1, math.ceil(bnorm[1] * math.pi * 2 / spacing))
    nkpz = max(1, math.ceil(bnorm[2] * math.pi * 2 / spacing))

    # Apply multiplier if provided
    if multipl:
        nkpx *= multipl[0]
        nkpy *= multipl[1]
        nkpz *= multipl[2]

    return nkpx, nkpy, nkpz

# Example usage
bnorm = [1.0, 2.0, 3.0]  # Example lattice norms
spacing = 0.5  # Example KPOINTS spacing
multipl = [2, 2, 2]  # Example multiplier

nkpx, nkpy, nkpz = calculate_grid_dimensions(bnorm, spacing, multipl)
print("Grid dimensions:", nkpx, nkpy, nkpz)

```
This Python function `calculate_grid_dimensions` mirrors the Fortran logic for determining the Monkhorst-Pack grid dimensions (`NKPX`, `NKPY`, `NKPZ`) based on the lattice norms (`BNORM`), the Monkhorst-Pack spacing (`KSPACING`), and an optional multiplier (`MULTIPL`).


>KSPACING = 0.02 seems very tight. i doubt you need that much

What are the optimal `KSPACING` values, based on your experiences?

See [here](https://www.vasp.at/forum/viewtopic.php?t=19459) for the related discussion.

**Reply by janosh** [reply]
[MP uses 0.22 for metals and 0.44 for insulators with smooth interpolation in between based on band gap](https://github.com/materialsproject/atomate2/blob/35e97176074a6b17f70c535fea44318822d39b59/src/atomate2/vasp/sets/base.py#L1106-L1115)

**Reply by hongyi-zhao** [reply]
> https://github.com/materialsproject/pymatgen/discussions/3709#discussioncomment-8902808
> KSPACING = 0.02 seems very tight. i doubt you need that much

After a second thought, the units we're talking about here may be different, but I'm talking roughly based on the units in vaspkit, aka, `2*PI/Angstrom`, as described [here](https://vaspkit.com/tutorials.html):

![image](https://github.com/materialsproject/pymatgen/assets/11155854/28c3e53a-9e0d-4dbf-b6f9-0920e1b1ba00)

But in vasp, the unit for [`KSPACING`](https://www.vasp.at/wiki/index.php/KSPACING) is `\AA ^ -1`. So, in my case, what I'm trying to say is `0.02 * 2 * PI` should be used to set `KSPACING`.

**Reply by hongyi-zhao** [reply]
See below for the related discussions:
https://github.com/JaGeo/LobsterPy/issues/303#issuecomment-2186655943
https://www.vasp.at/forum/viewtopic.php?t=19459
